### PR TITLE
Fix handling of OOD issues in optimizer

### LINKED ISSUE #4297: Implement better handling of OOD issues in optimizer
**Is your feature request related to a problem? Please describe.**

Currently, if the optimizer is used, it creates a new temporary segment with a copy of data from those segments, which are currently under optimization. 
This process requires sometimes significant amount of disk space and if it is not enough, current implementation might panic.

**Describe the solution you'd like**

In optimization process, we can check if there are enough disk space to continue indexing or not.
A good space to do it is https://github.com/qdrant/qdrant/blob/ccf7f1d24f37dcfe82dad85fa66fb91352bd7a21/lib/collection/src/collection_manager/optimizers/segment_optimizer.rs#L100

**Describe alternatives you've considered**
Indexed segment is potentially a plain structure. There are not too many places where we can't handle the OOD error.
Alternatively, we can check if we can actually handle OOD errors in this case, instead of checking for empty space each time.

**Additional context**

We created a test that simulates the scenario of disk overflow https://github.com/qdrant/qdrant/pull/4267

The solution for this issue should resolve this test scenario in the most general way reasonable.

### PR DESCRIPTION
**Pull request by the-johnwick**
fixes #4297
/claim #4297

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by the-johnwick** [comment]
@generall  waiting for your feedback whether it is going in right direction or not.

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 1]
```diff
@@ -143,6 +156,17 @@ pub trait SegmentOptimizer {
             .copied()
             .unwrap_or(0);
 
+        // Estimate required disk space
+        let required_space = maximal_vector_store_size_bytes;
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:160]: I would argue, that required_space should also include the paylaod size.

I propose to take sum of optimizing segments disk size as a reference instead.

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 2]
```diff
@@ -38,6 +39,18 @@ pub struct OptimizerThresholds {
     pub indexing_threshold_kb: usize,
 }
 
+fn available_disk_space(path: &Path) -> u64 {
+    let mut sys = System::new_all();
+    sys.refresh_disks_list();
+
+    for disk in sys.disks() {
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:46]: We are already using `fs4` for he similar purpose https://github.com/qdrant/qdrant/blob/1efec29f65c7569a3ae3b76bbd511410556c53c6/lib/collection/src/shards/local_shard/disk_usage_watcher.rs#L106

I would prefer to keep the consistency

**Review by generall — CHANGES_REQUESTED** [review]
Hey @the-johnwick, thanks for the PR! Overall I agree with the approach, but there are a couple of thing I would prefer to change. Please see the comments

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs (hunk 1 above)
Comment (the-johnwick) [reply lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:160]: Sure, but I am confused about calculating payload's total size.

**Comment by Rutik7066** [comment]
@generall  added the fs4 to check available space and the payload size to the required size. It may not be the right way to calculate payload size. Please let me know .

**Comment by the-johnwick** [comment]
@generall added the fs4 to check available space and the payload size to the required size. It may not be the right way to calculate payload size. Please let me know.

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+use std::borrow::Borrow;
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:1]: this doesn't look related, is it?

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 4]
```diff
@@ -128,11 +130,11 @@ pub trait SegmentOptimizer {
                 }
             };
             let locked_segment = segment.read();
-
             for vector_name in locked_segment.vector_names() {
                 let vector_size = locked_segment.available_vectors_size_in_bytes(&vector_name)?;
                 let size = bytes_count_by_vector_name.entry(vector_name).or_insert(0);
                 *size += vector_size;
+                total_payload_size_bytes += vector_size
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:137]: I don't think it is a correct way to estimate payload.

I think what we need to do is to simply check what's the current disk usage of the segment is, not try to estimate it with some form of proxy value.

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 5]
```diff
@@ -143,6 +145,21 @@ pub trait SegmentOptimizer {
...
             .unwrap_or(0);
 
+        let mut available_space;
+        match fs4::available_space(self.temp_path()) {
+            Some(size) => available_space = size,
+            Err(error) => CollectionError::service_error(format!(
+                "Failed to fetch available space for path: {} due to: {}",
+                self.temp_path().to_owned(),
+                error.to_string()
+            )),
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:155]: Some file systems might not support it, so we would prefer to go on with the optimization instead of failing. 
So logging the error should be enough

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs (hunk 3 above)
Comment (the-johnwick) [reply lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:1]: Yes, I think it is added by mistake. I will remove it.

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs (hunk 5 above)
Comment (the-johnwick) [reply lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:155]: Sure I will update the changes.

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 6]
```diff
@@ -101,6 +101,30 @@ pub trait SegmentOptimizer {
...
+        match fs4::available_space(self.temp_path()) {
+            Some(size) => available_space = size,
+            Err(error) => log::error!(
+                "Failed to fetch available space for path: {} due to: {}",
+                self.temp_path().to_owned(),
+                error.to_string(),
+            ),
+        }
+        let mut total_space;
+        match fs4::total_space(self.segments_path()) {
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:114]: it should be sum of `optimizing_segments`, not all segments in overall

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 7]
```diff
@@ -128,6 +128,14 @@ pub trait SegmentOptimizer {
                 }
             };
             let locked_segment = segment.read();
+            match fs4::total_space(locked_segment.data_path()) {
+                Some(size) => total_space = size,
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:132]:
```suggestion
                Some(size) => total_space += size,
```

File: lib/collection/src/collection_manager/optimizers/segment_optimizer.rs [hunk 8]
```diff
@@ -117,7 +117,7 @@ pub trait SegmentOptimizer {
         //     image_vectors: 10100 * dim * VECTOR_ELEMENT_SIZE
         // }
         let mut bytes_count_by_vector_name = HashMap::new();
-
+        let mut total_required_space;
```
Comment (generall) [inline lib/collection/src/collection_manager/optimizers/segment_optimizer.rs:120]:
```suggestion
        let mut total_required_space = 0;
```

**Comment by the-johnwick** [comment]
@generall  apologies for such silly mistakes.

**Comment by generall** [comment]
Hey @the-johnwick, could you please rebase on top of https://github.com/qdrant/qdrant/pull/4267
So we can be sure that the test case is fixed

**Comment by the-johnwick** [comment]
@generall  Sure.

File: Cargo.toml [hunk 9]
```diff
@@ -70,7 +70,7 @@ tokio = { workspace = true }
 
 actix-cors = "0.7.0"
 actix-files = "0.6.6"
-actix-web = { version = "4.7.0", optional = true, features = ["rustls-0_23", "actix-tls"] }
```
Comment (generall) [inline Cargo.toml:73]: how is this change related to the issue?

Comment (the-johnwick) [reply Cargo.toml:73]: No. actually I have messed up while trying to rebase

Comment (the-johnwick) [reply Cargo.toml:73]: Do you know, how to undo the changes?

**Comment by the-johnwick** [comment]
@generall Can we merge this.

**Comment by generall** [comment]
Close in favor of https://github.com/qdrant/qdrant/pull/4578
